File: algorithms/implementation/Counting Valleys/counting-valleys.js
/*
  Counting Valleys
*/
/*
  https://www.hackerrank.com/challenges/counting-valleys/problem
*/

function countingValleys(n, s) {
  let valleys = 0;
  let elevation = 0;

  for (let char of s) {
    if (char === 'U') elevation += 1;
    else elevation -= 1;

    if (!elevation && char === 'U') valleys += 1;
  }
  return valleys;
}

File: algorithms/implementation/Counting Valleys/counting-valleys.py
# Counting Valleys
# https://www.hackerrank.com/challenges/counting-valleys/problem

def countingValleys(n, s):
  valleys = 0
  elevation = 0

  for char in s:
    if char == 'U':
      elevation += 1
    else:
      elevation -= 1
    
    if (elevation == 0 and char == 'U'):
      valleys += 1
  
  return valleys
